### DISCUSSION TITLE
How to use timestamp

### DISCUSSION THREAD

**Discussion by hmmftg** [sijms/go-ora#126] [Q&A] [answered]
Hello
I'm using `date`  as out parameter and it's working well
But when I add a `timestamp` out parameter (as `time.Time` type) to my procedure, the  nano second part is missing when I print output:

`var timeStamp time.Time`
`_, err = db1.Exec("BEGIN GET_TIME(:1); END;", sql.Out{Dest: &timeStamp})`
`if err != nil {`
`    log.Fatal(err)`
`}`
`fmt.Println(timeStamp.Format("2006-01-02T15:04:05.999999999Z07:00"))`

output:  `2021-10-23 08:10:13`

**Comment by sijms** [comment]
this is the function I use:
```Golang
func TestTimestamp(conn *sql.DB) error {
	row := conn.QueryRow("SELECT TO_TIMESTAMP ('10-Sep-02 14:10:10.123456', 'DD-Mon-RR HH24:MI:SS.FF')FROM DUAL")
	var date time.Time
	if row != nil {
		if row.Err() != nil {
			return row.Err()
		}
		err := row.Scan(&date)
		if err != nil {
			return err
		}
		fmt.Println(date)
	}
	return nil
}
```
and this is the result
`2002-09-10 14:10:10.123456 +0000 UTC`

**Reply by sijms** [reply]
Sorry I see you use output parameter I will test with output pars and inform you

**Reply by sijms** [reply]
I find the issue
when you pass a time.Time parameter I suspect that it is date thus I add the following to the parameter
```Golang
case *time.Time:
	param.Value = val
	param.BValue = converters.EncodeDate(*val)
	param.DataType = DATE
	param.ContFlag = 0
	param.MaxLen = 11
	param.MaxCharLen = 11
```
if I change DATE to TIMESTAMP It work
the solution I see is adding a new type called TimeStamp to the package
this is the code:
```Golang
var date go_ora.TimeStamp
_, err := conn.Exec(`BEGIN 
        SELECT TO_TIMESTAMP ('10-Sep-02 14:10:10.123456789', 'DD-Mon-RR HH24:MI:SS.FF') into :1 FROM DUAL; 
    END;`, sql.Out{Dest: &date})
if err != nil {
	return err
}
fmt.Println(time.Time(date))
```
if you have another solution I will be happy if you share with me

**Reply by hmmftg** [reply]
Hi
thanks for your reply
I also think separating date and timestamp is better than mixing them
and as most plsql developers use date, it's better to consider time struct as date

**Comment by sijms** [comment, marked as answer]
this is the code:
```Golang
var date go_ora.TimeStamp
_, err := conn.Exec(`BEGIN 
        SELECT TO_TIMESTAMP ('10-Sep-02 14:10:10.123456789', 'DD-Mon-RR HH24:MI:SS.FF') into :1 FROM DUAL; 
    END;`, sql.Out{Dest: &date})
if err != nil {
	return err
}
fmt.Println(time.Time(date))
```

**Reply by hmmftg** [reply]
Hi
Please add a new pull request(I dont know how does this work!) for adding new type TimeStamp to the package  
I can also help with code if you like, because I shall be so happy to make this project better
thanks

**Reply by sijms** [reply]
I think you need to make it from you account. use the following [help](https://opensource.com/article/19/7/create-pull-request-github)
